src/mongo_odm/util.hpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <chrono>
+#include <ctime>
+#include <string>
+#include <type_traits>
+
+#include <bsoncxx/stdx/optional.hpp>
+#include <bsoncxx/types.hpp>
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+template <bool...>
+struct bool_pack;
+
+/**
+ * A templated struct for determining whether a variadic list of boolean conditions is all true.
+ * i.e. a logical AND of a set of conditions.
+ */
+template <bool... bs>
+struct all_true : public std::is_same<bool_pack<bs..., true>, bool_pack<true, bs...>> {};
```

Comment:
You're doing enough fancy stuff here that I think it might be worth it to look at some of the C++17 type traits that work on variadics:

http://en.cppreference.com/w/cpp/types/conjunction
http://en.cppreference.com/w/cpp/types/disjunction